### ISSUE TITLE
Unhandled Exception: FormatException: Unrecognized string escape

### ISSUE THREAD

**Issue by royabousamra** [shamblett/xml2json#29]
Happening to a phrase like "04/04/2019$$##123333%%##*\\+|=\€€^{€€\\€\\>_€_€". The first error indicates it's a problem with the vertical bar | but a modification to the previous phrase, "04/04/2019$$##+|=\\€", shows otherwise. 

Error for original phrase:

> E/flutter (27723): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: FormatException: Unrecognized string escape (at character 414)
> E/flutter (27723): ..."04/04/2019$$##laaalaa 123333%%##*\\+|=\€€^{€€\\€\\>_€_€", "workingState": ...
> E/flutter (27723):  ........................................................................................^
> E/flutter (27723): 
> E/flutter (27723): #0      _ChunkedJsonParser.fail (dart:convert-patch/convert_patch.dart:1404:5)
> E/flutter (27723): #1      _ChunkedJsonParser.parseStringEscape (dart:convert-patch/convert_patch.dart:1184:9)
> E/flutter (27723): #2      _ChunkedJsonParser.parseStringToBuffer (dart:convert-patch/convert_patch.dart:1123:18)
> E/flutter (27723): #3      _ChunkedJsonParser.parseString (dart:convert-patch/convert_patch.dart:1038:16)
> E/flutter (27723): #4      _ChunkedJsonParser.parse (dart:convert-patch/convert_patch.dart:859:22)
> E/flutter (27723): #5      _parseJson (dart:convert-patch/convert_patch.dart:40:10)

Error for modified phrase:

> E/flutter (27723): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: FormatException: Unrecognized string escape (at character 398)
> E/flutter (27723): ...loyeeFeedback": "04/04/2019$$##+|=\\€", "workingState": ...
> E/flutter (27723): ....................................................................................^
> E/flutter (27723): 
> E/flutter (27723): #0      _ChunkedJsonParser.fail (dart:convert-patch/convert_patch.dart:1404:5)
> E/flutter (27723): #1      _ChunkedJsonParser.parseStringEscape (dart:convert-patch/convert_patch.dart:1184:9)
> E/flutter (27723): #2      _ChunkedJsonParser.parseStringToBuffer (dart:convert-patch/convert_patch.dart:1123:18)
> E/flutter (27723): #3      _ChunkedJsonParser.parseString (dart:convert-patch/convert_patch.dart:1038:16)
> E/flutter (27723): #4      _ChunkedJsonParser.parse (dart:convert-patch/convert_patch.dart:859:22)
> E/flutter (27723): #5      _parseJson (dart:convert-patch/convert_patch.dart:40:10)

Using version 4.4.0

**Comment by shamblett** [comment]
Your stack traces show an error in the dart:convert _parseJson library, not the xml2json package. Please supply your input XML and which conversion you are using, its possible the package is preprocessing the XML string incorrectly.

**Comment by royabousamra** [comment]
XML input that's causing the unhandled exception error:

> <?xml version="1.0" encoding="UTF-8"?>
> <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
>    <soapenv:Body>
>       <ns1:GetCommentsResponse xmlns:ns1="http://website.com/AppWS/">
>          <status>true</status>
>          <message>Succeed</message>
>          <comments>
>             <feedback>04/04/2019\$\$##+|=\€</feedback>
>          </comments>
>       </ns1:GetCommentsResponse >
>    </soapenv:Body>
> </soapenv:Envelope>

For conversion from XML to JSON, toParker() is being used:

> final xmlParser = Xml2Json();
> xmlParser.parse(response.body);
> jsonResponse = xmlParser.toParker();

**Comment by shamblett** [comment]
Interestingly, on the VM, this test using the data supplied above  -
```
import 'dart:convert';
import 'package:pana/models.dart';
import 'package:test/test.dart';
import 'package:xml2json/xml2json.dart';

void main() {
  test(
      'Unhandled Exception: FormatException: Unrecognized string escape',
      () {
        const input = '''<?xml version="1.0" encoding="UTF-8"?>
            <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
        <soapenv:Body>
        <ns1:GetCommentsResponse xmlns:ns1="http://website.com/AppWS/">
        <status>true</status>
        <message>Succeed</message>
        <comments>
        <feedback>04/04/2019\$\$##+|=\€</feedback>
        </comments>
        </ns1:GetCommentsResponse >
        </soapenv:Body>
        </soapenv:Envelope>''';

        final xmlParser = Xml2Json();
        xmlParser.parse(input);
        var jsonResponse = xmlParser.toParker();
        print(jsonResponse);
  });
}
```

produces this output -
```
{"soapenv:Envelope": {"soapenv:Body": {"ns1:GetCommentsResponse": {"status": "true", "message": "Succeed", "comments": {"feedback": "04/04/2019$$##+|=€"}}}}}
```

so I'm not seeing the error you are seeing on flutter.

**Comment by royabousamra** [comment]
Sorry, can you try the following input:

```
const input = <?xml version="1.0" encoding="UTF-8"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
   <soapenv:Body>
      <ns1:GetCommentsResponse xmlns:ns1="http://website.com/AppWS/">
         <status>true</status>
         <message>Succeed</message>
         <comments>
            <feedback>04/04/2019\$\$##+|=\\€</feedback>
         </comments>
      </ns1:GetCommentsResponse >
   </soapenv:Body>
</soapenv:Envelope>
```

**Comment by shamblett** [comment]
Gives me -
```
{"soapenv:Envelope": {"soapenv:Body": {"ns1:GetCommentsResponse": {"status": "true", "message": "Succeed", "comments": {"feedback": "04/04/2019$$##+|=\€"}}}}}
```

**Comment by royabousamra** [comment]
Validating the above outputted JSON response shows a parse error (checked here: https://jsonlint.com/). The fix is "escaping unescaped character" (checked here: https://jsonformatter.curiousconcept.com/#).

**Comment by shamblett** [comment]
OK, the problem with this is windows file paths, some users encode these thus -
```
  const input = '''<?xml version="1.0" encoding="UTF-8"?>
            <path>C:work\\fred\\joe</path>''';

```
This ends up being  '{"path": "C:work\fred\joe"}' which fails the json linter but is nevertheless what they expect.
Which gives the problem of how does the package know whether '\\' is intended or not?

**Comment by simonkimi** [comment]
Your input is `<feedback>04/04/2019\$\$##+|=\€</feedback>`, Does the "\\" have an escaped meaning?
In the XML specification, only "&","<",">" needs to be escaped, so the output should be expected
```
const input = r'''<?xml version="1.0" encoding="UTF-8"?>
            <feedback>04/04/2019\$\$##+|=\€</feedback>''';

final xmlParser = Xml2Json();
xmlParser.parse(input);
var jsonResponse = xmlParser.toParker();
print('raw json $jsonResponse}');
print('decoded json: ${jsonDecode(jsonResponse)}');
```
output:
```
raw json {"feedback": "04/04/2019\\$\\$##+|=\\€"}}
decoded json: {feedback: 04/04/2019\$\$##+|=\€}
```

**Comment by simonkimi** [comment]
> 
> 
> Interestingly, on the VM, this test using the data supplied above -
> 
> ```
> import 'dart:convert';
> import 'package:pana/models.dart';
> import 'package:test/test.dart';
> import 'package:xml2json/xml2json.dart';
> 
> void main() {
>   test(
>       'Unhandled Exception: FormatException: Unrecognized string escape',
>       () {
>         const input = '''<?xml version="1.0" encoding="UTF-8"?>
>             <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
>         <soapenv:Body>
>         <ns1:GetCommentsResponse xmlns:ns1="http://website.com/AppWS/">
>         <status>true</status>
>         <message>Succeed</message>
>         <comments>
>         <feedback>04/04/2019\$\$##+|=\€</feedback>
>         </comments>
>         </ns1:GetCommentsResponse >
>         </soapenv:Body>
>         </soapenv:Envelope>''';
> 
>         final xmlParser = Xml2Json();
>         xmlParser.parse(input);
>         var jsonResponse = xmlParser.toParker();
>         print(jsonResponse);
>   });
> }
> ```
> 
> produces this output -
> 
> ```
> {"soapenv:Envelope": {"soapenv:Body": {"ns1:GetCommentsResponse": {"status": "true", "message": "Succeed", "comments": {"feedback": "04/04/2019$$##+|=€"}}}}}
> ```
> 
> so I'm not seeing the error you are seeing on flutter.

In this test, you might make a mistake, the raw xml is `<feedback>04/04/2019\$\$##+|=\€</feedback>`, In Dart, however, `\$`is escaped as `$`, `\any` is escaped as `any`. In other words, all the `\` are lost, the original meaning of XML has changed.
To avoid this situation, you should add "r" before the string to avoid escape, or write XML to a file and use File() to read it.

```dart
void main() {
  test('dart escape', () {
    final raw = r"""<feedback>04/04/2019\$\$##+|=\€</feedback>""";
    final escaped = """<feedback>04/04/2019\$\$##+|=\€</feedback>""";
    expect(raw, escaped);
  });
}
```
```
Expected: '<feedback>04/04/2019$$##+|=€</feedback>'
  Actual: '<feedback>04/04/2019\\$\\$##+|=\\€</feedback>'
   Which: is different.
          Expected: ... 04/04/2019$$##+|=€</ ...
            Actual: ... 04/04/2019\\$\\$##+| ...
                                  ^
           Differ at offset 20
```

**Comment by shamblett** [comment]
I hear what you are saying but the package gets its strings from the xml package, not from anywhere you can put an 'r' in front of it, which only seems to be a source code construct, writing to a file besides being overkill wont' help in the browser so we can rule this out.

I think all I can endeavor to do is not generate invalid json, we shouldn't do this, if this breaks things for some users then so be it.

**Comment by shamblett** [comment]
Should be OK now, we should not now be generating invalid json, package released at version 5.0.0

**Comment by royabousamra** [comment]
Fix works great, thank you.
